// ripple-command/config.js
'use strict';

var _ = require('lodash');

const defaults = {
    server: 'wss://s1.ripple.com',
    offline: false,
    logging: true,
    loglevel: 'warn',
    logfile: 'ripple-command.log',
};

const env_mapping = {
    server: 'RIPPLE_COMMAND_SERVER',
    logging: 'RIPPLE_COMMAND_LOGGING',
    loglevel: 'RIPPLE_COMMAND_LOGLEVEL',
    logfile: 'RIPPLE_COMMAND_LOGFILE',
};

/**
 * Simple utility for handling configuration setting across the application.
 */
class Config {

    /**
     * Fill in the first config with the default values defined above.
     *
     * Then, process each env_mapping to set config values based on env values.
     */
    constructor() {
        this.config = _.extend({}, defaults);
        _.each(env_mapping, (env_var, config_key) => {
            if (process.env[env_var] !== undefined) {
                this.config[config_key] = process.env[env_var];
            }
        });
    }

    /**
     * Get a value from the current config.
     *
     * @param  {string} name
     * @return {mixed}
     */
    get(name) {
        return this.config[name];
    }

    /**
     * Get a default value, as defined in the defaults constant.
     *
     * @param  {string} name
     * @return {mixed}
     */
    getDefault(name) {
        return defaults[name];
    }

    /**
     * Set a given config value.
     *
     * @param {string} name
     * @param {mixed} value
     */
    set(name, value) {
        this.config[name] = value;
    }

    /**
     * Absorb any number of object-based configurations into
     * the current configuration.
     *
     * Each variadic argument should be an object representing a config setting
     * in the format of key:value
     *
     * Undefined values will be ignored.
     *
     * @param  {...[object, ...]} options
     */
    absorb(...options) {
        options = options.map((obj, key) => {
            return _.pickBy(obj, (v, k) => {
                return v !== undefined;
            });
        });
        options.unshift(this.config);
        this.config = _.extend.apply(null, options);
    }

}

module.exports = {
    Config: new Config(),
};
